lib/web_ui/lib/src/engine/util.dart

```diff
@@ -54,115 +54,99 @@ String matrix4ToCssTransform(Matrix4 matrix) {
-  final bool isHighDevicePixelRatioScreen =
-      EngineWindow.browserDevicePixelRatio > 1.0;
-
-  if (transformKind == TransformKind.scaleAndTranslate2d) {
-    final String cssTransform = float64ListToCssTransform2d(matrix4);
-    element.style
-      ..transformOrigin = '0 0 0'
-      ..transform = cssTransform
-      ..top = null
-      ..left = null;
-  } else if (transformKind == TransformKind.complex || isHighDevicePixelRatioScreen) {
-    final String cssTransform = float64ListToCssTransform3d(matrix4);
-    element.style
-      ..transformOrigin = '0 0 0'
-      ..transform = cssTransform
-      ..top = null
-      ..left = null;
-  } else if (transformKind == TransformKind.translation2d) {
-    final double ty = matrix4[13];
-    final double tx = matrix4[12];
-    element.style
-      ..transformOrigin = null
-      ..transform = null
-      ..left = '${tx}px'
-      ..top = '${ty}px';
+String float64ListToCssTransform(Float64List matrix) {
+  assert(matrix.length == 16);
+  final TransformKind transformKind = transformKindOf(matrix);
+  if (transformKind == TransformKind.transform2d) {
+    return float64ListToCssTransform2d(matrix);
+  } else if (transformKind == TransformKind.complex) {
+    return float64ListToCssTransform3d(matrix);
   } else {
     assert(transformKind == TransformKind.identity);
-    element.style
-      ..transformOrigin = null
-      ..transform = null
-      ..left = null
-      ..top = null;
+    return null;
   }
 }
 
 /// The kind of effect a transform matrix performs.
 enum TransformKind {
   /// No effect.
+  ///
+  /// We do not want to set any CSS properties in this case.
   identity,
 
-  /// A transform that contains only 2d scale and transform.
-  scaleAndTranslate2d,
-
-  /// A translation along either X or Y axes, or both.
-  translation2d,
+  /// A transform that contains only 2d scale, rotation, and translation.
+  ///
+  /// We prefer to use "matrix" instead of "matrix3d" in this case.
+  transform2d,
 
   /// All other kinds of transforms.
+  ///
+  /// In this case we will use "matrix3d".
   complex,
 }
 
 /// Detects the kind of transform the [matrix] performs.
 TransformKind transformKindOf(Float64List matrix) {
   assert(matrix.length == 16);
   final Float64List m = matrix;
-  final double ty = m[13];
-  final double tx = m[12];
 
   // If matrix contains scaling, rotation, z translation or
   // perspective transform, it is not considered simple.
   final bool isSimple2dTransform =
-      // m[0] - scale x is simple
-      m[1] == 0.0 &&
-      m[2] == 0.0 &&
-      m[3] == 0.0 &&
-      m[4] == 0.0 &&
-      // m[5] - scale y is simple
-      m[6] == 0.0 &&
-      m[7] == 0.0 &&
-      m[8] == 0.0 &&
-      m[9] == 0.0 &&
-      m[10] == 1.0 &&
-      m[11] == 0.0 &&
-      // m[12] - x translation is simple
-      // m[13] - y translation is simple
+      m[15] == 1.0 &&  // start reading from the last element to eliminate range checks in subsequent reads.
```

Comment:
Wow, interesting. Is this a dart2js thing?

Reply:
I've heard this about many compilers, C++, Java, Dart, etc (more info: https://en.wikipedia.org/wiki/Bounds-checking_elimination).